onnx/onnxifi_loader.c

```diff
@@ -0,0 +1,241 @@
+#ifdef _WIN32
+#include <windows.h>
+#include <malloc.h> /* for _alloca */
+#else
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <alloca.h>
+#include <dlfcn.h>
+#endif
+
+#include <onnxifi_loader.h>
+
+/* ONNXIFI_LOADER_LOGGING macro enables/disables logging. Its OFF by default. */
+#ifndef ONNXIFI_LOADER_LOGGING
+#define ONNXIFI_LOADER_LOGGING 0
+#endif
+
+#if ONNXIFI_LOADER_LOGGING
+#if defined(__ANDROID__)
+#include <android/log.h>
+/* Tag used for logging on Android */
+#define ONNXIFI_LOADER_ANDROID_LOG_TAG "ONNX-LOADER"
+#else
+#include <stdio.h>
+#endif
+#endif
+
+#if defined(__APPLE__)
+#define ONNXIFI_LIBRARY_NAME "libonnxifi.dylib"
+#elif defined(_WIN32)
+#define ONNXIFI_LIBRARY_NAME L"onnxifi.dll"
+#else
+#define ONNXIFI_LIBRARY_NAME "libonnxifi.so"
+#endif
+
+/* Order must match declaration order in onnxifi_library structure */
+static const char onnxifi_function_names[] =
+    "onnxGetBackendIDs\0"
+    "onnxReleaseBackendID\0"
+    "onnxGetBackendInfo\0"
+    "onnxGetBackendCompatibility\0"
+    "onnxInitBackend\0"
+    "onnxReleaseBackend\0"
+    "onnxInitEvent\0"
+    "onnxSignalEvent\0"
+    "onnxWaitEvent\0"
+    "onnxReleaseEvent\0"
+    "onnxInitGraph\0"
+    "onnxSetGraphIO\0"
+    "onnxRunGraph\0"
+    "onnxReleaseGraph\0";
+
+/* Length of the longest function, including terminating null */
+#define ONNXIFI_FUNCTION_NAME_MAX sizeof("onnxGetBackendCompatibility")
+
+int ONNXIFI_ABI onnxifi_load(
+  struct onnxifi_library* onnx,
+  uint32_t flags,
+#ifdef _WIN32
+  const wchar_t* path,
+#else
+  const char* path,
+#endif
+  const char* suffix)
+{
+  size_t i;
+  const char* function_name;
+  char* buffer;
+  size_t buffer_length;
+  size_t suffix_length;
+#ifdef _WIN32
+  LPCSTR format_arguments[2];
+#endif
+
+  if (onnx == NULL) {
+    return 0;
+  }
+
+#ifdef _WIN32
+  ZeroMemory(onnx, sizeof(struct onnxifi_library));
+#else
+  memset(onnx, 0, sizeof(struct onnxifi_library));
+#endif
+  if (!(flags & ONNXIFI_LOADER_FLAG_VERSION_1_0)) {
+    /* Unknown ONNXIFI version requested */
+    return 0;
+  }
+
+  if (path == NULL) {
+    path = ONNXIFI_LIBRARY_NAME;
+  }
+  if (suffix == NULL) {
+    suffix = "";
+  }
+
+#ifdef _WIN32
+  buffer_length = ONNXIFI_FUNCTION_NAME_MAX + lstrlenA(suffix);
+  buffer = (char*) _alloca(buffer_length);
+#else
+  buffer_length = ONNXIFI_FUNCTION_NAME_MAX + strlen(suffix);
+  buffer = (char*) alloca(buffer_length);
+#endif
+
+#ifdef _WIN32
```

Comment:
Not sure it will be helpful, as this sequence is specific to dynamic loading. For other functionality, it is customary to use `SetLastError`/`errno` combination.